userapi/src/main/java/fk/prof/userapi/util/StackLineParser.java

```diff
@@ -0,0 +1,122 @@
+package fk.prof.userapi.util;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** A utility class which parses and converts Java names from their corresponding JVM Types signatures
+ - from string of format className#methodName (argTypes)retType
+ * Created by rohit.patiyal on 26/04/17.
+ */
+public class StackLineParser {
+    private static Map<Character, String> signToPrimitiveMap = new HashMap<Character, String>(){
+        {
+            put('V',"void");
+            put('Z',"boolean");
+            put('B',"byte");
+            put('C',"char");
+            put('S',"short");
+            put('I',"int");
+            put('J',"long");
+            put('F',"float");
+            put('D',"double");
+        }
+    };
+
+    private static List<JVMTypeSignatureParser> parsers = new ArrayList<JVMTypeSignatureParser>(){
+        {
+            add(StackLineParser::parseTerminalType);
+            add(StackLineParser::parseMethodName);
+            add(StackLineParser::parseArgs);
+            add(StackLineParser::parseRet);
+        }
+    } ;
+
+    public static String convertJVMTypeSignToJava(String text){
+        StringBuilder sb = new StringBuilder();
+        int textIndex = 0;
+        int sbLength = 0;
+        text = text.replace('/','.');
+        for(JVMTypeSignatureParser parser : parsers){
+            int res = parser.parse(sb, text, textIndex);
+            if(res == -1){
+                return sb.substring(0, sbLength) + text.substring(textIndex);
+            }
+            textIndex = res;
+            sbLength = sb.length();
+        }
+        return sb.toString();
+    }
+
+
+
+    @FunctionalInterface
+    private interface JVMTypeSignatureParser{
+        int parse(StringBuilder sb, String text, int textIndex);
+    }
+
+    private static int parseTerminalType(StringBuilder sb, String text, int textIndex){
```

Comment:
The name terminal was to suggest a symbol which couldn't expand by substituting other symbols. But will change this, given that the array case actually has further substitutions. Hence our definition of Type(className/Primitive type/Array of these) does not make it a pure terminal symbol.